include/object-with-language.php

```diff
@@ -0,0 +1,337 @@
+<?php
+/**
+ * @package Polylang
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Abstract class to use for object types that support at least one language.
+ *
+ * @since 3.4
+ */
+abstract class PLL_Object_With_Language implements PLL_Object_With_Language_Interface {
+
+	/**
+	 * @var PLL_Model
+	 */
+	public $model;
+
+	/**
+	 * Taxonomy name for the languages.
+	 *
+	 * @var string
+	 *
+	 * @phpstan-var non-empty-string
+	 */
+	protected $tax_language;
+
+	/**
+	 * Object type to use when registering the taxonomy.
+	 * Left empty for object types that are not a taxonomy.
+	 *
+	 * @var string|null
+	 *
+	 * @phpstan-var non-empty-string|null
+	 */
+	protected $object_type = null;
+
+	/**
+	 * Name of the `PLL_Language` property that stores the term_taxonomy ID.
+	 *
+	 * @var string
+	 *
+	 * @phpstan-var non-empty-string
+	 */
+	protected $tax_tt_prop_name;
+
+	/**
+	 * List of taxonomies to cache in get_object_term().
+	 *
+	 * @var string[]
+	 * @see PLL_Object_With_Language::get_object_term()
+	 */
+	protected $tax_to_cache = array();
+
+	/**
+	 * Default alias corresponding to the object's DB table.
+	 *
+	 * @var string
+	 * @see PLL_Object_With_Language::join_clause()
+	 *
+	 * @phpstan-var non-empty-string
+	 */
+	protected $db_default_alias;
+
+	/**
+	 * Name of the DB column containing the object's ID.
+	 *
+	 * @var string
+	 * @see PLL_Object_With_Language::join_clause()
+	 *
+	 * @phpstan-var non-empty-string
+	 */
+	protected $db_id_column;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 3.4
+	 *
+	 * @param PLL_Model $model Instance of `PLL_Model`.
+	 */
+	public function __construct( PLL_Model &$model ) {
+		$this->model          = &$model;
+		$this->tax_to_cache[] = $this->tax_language;
+
+		/*
+		 * Register our taxonomy as soon as possible.
+		 * This is early registration, not ready for rewrite rules as $wp_rewrite will be setup later.
+		 */
+		register_taxonomy(
+			$this->tax_language,
+			(array) $this->object_type,
+			array(
+				'label'     => false,
+				'public'    => false,
+				'query_var' => false,
+				'rewrite'   => false,
+				'_pll'      => true,
+			)
+		);
+	}
+
+	/**
+	 * Returns the language taxonomy name.
+	 *
+	 * @since 3.4
+	 *
+	 * @return string
+	 *
+	 * @phpstan-return non-empty-string
+	 */
+	public function get_tax_language() {
+		return $this->tax_language;
+	}
+
+	/**
+	 * Add hooks.
+	 *
+	 * @since 3.4
+	 *
+	 * @return self
+	 */
+	public function init() {
+		return $this;
+	}
```

Comment:
Forget.

Reply:
For the record:
- Hooks are removed from the constructor to this method, so an instance can be created without launching the hooks.
- This method is created in this abstract class, so classes that extend this class don't have to create this method if they don't need it.
- For backward compatibility and preventing us to edit a lot of files (especially in tests), this method is called in `PLL_Translated_Post`'s and `PLL_Translated_Term`'s constructor. This may happen in the future though.
- New classes that extend this one, like the one coming to Polylang for WooCommerce, can call this method outside of the constructor.